### ISSUE TITLE
The Parallel.ForEachAsync with synchronous source can block many threads

### ISSUE THREAD

**Issue by theodorzoulias** [dotnet/runtime#81877]
### Description

I am using the [`Parallel.ForEachAsync`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.parallel.foreachasync) method with a `BlockingCollection<T>` as source, in a producer-consumer setup, and I observed a high number of threads used by the parallel loop. The `body` of the loop is completely asynchronous. I expected that the loop will use a small number of threads, but actually the thread usage goes up to the value of the configured `MaxDegreeOfParallelism`. This behavior is observed only with a synchronous source. If I convert the source to an `IAsyncEnumerable<T>` (using a converter from the `System.Linq.Async` package), the thread utilization is low and according to my expectations.

### Reproduction Steps

Here is a minimal demo:

```C#
BlockingCollection<int> queue = new();

Thread producer = new Thread(() =>
{
    for (int i = 1; i <= 50; i++)
    {
        Thread.Sleep(50); // Simulate a blocking I/O operation.
        queue.Add(i);
    }
    queue.CompleteAdding();
});
producer.Start();

ThreadPool.SetMinThreads(workerThreads: 10, completionPortThreads: 10);
ParallelOptions options = new() { MaxDegreeOfParallelism = 10 };

IEnumerable<int> source = queue.GetConsumingEnumerable().AsEnumerable();
//IAsyncEnumerable<int> source = queue.GetConsumingEnumerable().ToAsyncEnumerable();

await Parallel.ForEachAsync(source, options, async (x, _) =>
{
    Console.WriteLine($"Processing #{x}, ThreadCount: {ThreadPool.ThreadCount}");
    await Task.Delay(100); // Simulate an asynchronous I/O operation.
});
```

Output:

```none
Processing #1, ThreadCount: 3
Processing #2, ThreadCount: 3
Processing #3, ThreadCount: 3
Processing #4, ThreadCount: 3
Processing #5, ThreadCount: 4
...
Processing #45, ThreadCount: 10
Processing #46, ThreadCount: 10
Processing #47, ThreadCount: 10
Processing #48, ThreadCount: 10
Processing #49, ThreadCount: 10
Processing #50, ThreadCount: 10
```

[Online demo](https://dotnetfiddle.net/tCnwYy).

If I comment out the `.AsEnumerable()` line and uncomment the `.ToAsyncEnumerable()`, the output becomes like this:

```none
Processing #1, ThreadCount: 3
Processing #2, ThreadCount: 3
Processing #3, ThreadCount: 3
Processing #4, ThreadCount: 3
Processing #5, ThreadCount: 3
...
Processing #45, ThreadCount: 4
Processing #46, ThreadCount: 4
Processing #47, ThreadCount: 4
Processing #48, ThreadCount: 4
Processing #49, ThreadCount: 4
Processing #50, ThreadCount: 4
```

### Expected behavior

The [`ThreadPool.ThreadCount`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.threadpool.threadcount) has a low value during the parallel loop.

### Actual behavior

The `ThreadPool.ThreadCount` reaches the maximum value.

### Regression?

The behavior is the same on both .NET 6.0 and .NET 7.0.

### Known Workarounds

Convert the `source` to `IAsyncEnumerable<T>`, using the `ToAsyncEnumerable` extension method from the [`System.Linq.Async`](https://www.nuget.org/packages/System.Linq.Async) package.

### Configuration

- .NET 7.0
- Console application
- 64-bit operating system, x64-based processor

### Other information

I know that I could use a `Channel<T>` instead of a `BlockingCollection<T>` in my setup, and not have this problem. Nevertheless I thought that I should report this observation, because it is possible that the `Parallel.ForEachAsync` overload with synchronous `source` could be improved.

**Comment by stephentoub** [comment]
This is the nature of the API / by design.  It's doing what you asked it to do.

**Comment by theodorzoulias** [comment]
@stephentoub my expectation was that the API would not use more threads, than the threads required for invoking the `body` delegate. Assuming that the `body` is fully asynchronous, the number of threads should be lowish. I didn't expect that the API would saturate the `ThreadPool`, although I gave it a purely asynchronous work to do! Otherwise I could just revert to the synchronous `Parallel.ForEach`, and block the workers with `.GetAwaiter().GetResut()`.

Technically I think that using a `SemaphoreSlim(1, 1)` instead of locking is probably the simplest way to improve this behavior. Another idea could be to decommission workers who can't enter the `lock` immediately, and spawn new workers later.

**Comment by stephentoub** [comment]
> I didn't expect that the API would saturate the ThreadPool, although I gave it a purely asynchronous work to do!

But you didn't give it purely asynchronous work to do... you gave it synchronous work as part of the source enumerable.

> Technically I think that using a SemaphoreSlim(1, 1) instead of locking is probably the simplest way to improve this behavior.

It is, and out of all options it's also likely the most efficient... but compared to just taking a lock, it's also much less efficient than just using a lock around the enumerable.  To put it in perspective, I tried making the change, and I benchmarked:
```C#
[Benchmark]
public async Task ForEachAsync()
{
    await Parallel.ForEachAsync(Enumerable.Range(0, 1_000_000), (_, __) => ValueTask.CompletedTask);
}
```
The change to use SemaphoreSlim was 35x slower than what's currently being done with Monitor.  Even when changing the test to be:
```C#
[Benchmark]
public async Task ForEachAsync()
{
    await Parallel.ForEachAsync(Enumerable.Range(0, 1_000_000), async (_, __) => await Task.Yield());
}
```
it was still 9x slower.

The typical case is a source that doesn't block and that doesn't take a long time to MoveNext.  I'm not convinced it's worth changing the synchronization strategy to optimize for the case where it does at the expense of a measurable impact on the case where it doesn't.

**Comment by theodorzoulias** [comment]
People are using the `Parallel.ForEachAsync` to invoke asynchronous APIs, for doing I/O-bound work. For doing computations they can use the `Parallel.ForEach` API (which they also use frequently for I/O-bound work too, but this is another story). So returning a completed `ValueTask`, or the do-nothing `YieldAwaitable`, is unlikely to be a representative use of the API! Most real-world scenarios will invoke APIs with at least 1 millisecond latency, in which case the performance difference between `lock` and `SemaphoreSlim` most likely will not be measurable.

**Comment by stephentoub** [comment]
> Most real-world scenarios will invoke APIs with at least 1 millisecond latency

What are you basing this on?  Please share your source of data.

**Comment by theodorzoulias** [comment]
I am talking from my experience with StackOverflow, where I spend half of my day answering questions. This gives me some idea about how people use the .NET APIs, that are related to the tags that I am watching (concurrency, asynchronous, parallel-processing etc). I don't know if this counts as source of data, but that's what I have. :-)

**Comment by theodorzoulias** [comment]
Btw I just encountered a legitimate use case for feeding a `Parallel.ForEachAsync` loop with a `BlockingCollection<T>`: hierarchical parallel tree traversal. It is much more memory-efficient to traverse a tree using a stack for storing the descendent nodes, compared to using a queue, and there is no `Channel<T>` implementation with stack behavior (at least not in the standard .NET libraries). So out of necessity one would have to use a `BlockingCollection<T>` initialized with a `ConcurrentStack<T>`.

**Comment by theodorzoulias** [comment]
Another idea for improving the behavior of the `Parallel.ForEachAsync` could be to suspend the workers that can't enter the lock immediately, and resume one of the suspended workers each time an active worker returns from the [`state.Enumerator.MoveNext()`](https://github.com/dotnet/runtime/blob/8a96ee761d457fc8f0fcb3b1727151e34d9a2682/src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs#L100). I made some experiments, and it seems that this approach is about 5 times faster than the `SemaphoreSlim(1, 1)`, and about 5 times slower than the simple use of `lock`.

The other idea that I mentioned in a previous post, discarding the workers that can't acquire the lock immediately and recreating them later, is even faster, about 2,5 times slower than the `lock`, but it creates a lot of garbage.

Both of these approaches prevent the saturation of the `ThreadPool`, when the `source` sequence has blocking behavior.

**Comment by stephentoub** [comment]
> Another idea for improving the behavior of the Parallel.ForEachAsync could be to suspend the workers that can't enter the lock immediately, and resume one of the suspended workers each time an active worker returns from the [state.Enumerator.MoveNext()](https://github.com/dotnet/runtime/blob/8a96ee761d457fc8f0fcb3b1727151e34d9a2682/src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs#L100)

How is that logically different from await'ing semaphore.WaitAsync()?  Functionally isn't it identical?

**Comment by theodorzoulias** [comment]
After awaiting the `semaphore.WaitAsync()`, the semaphore has been acquired, but the worker who owns it must resume asynchronously on a different thread, which takes some time. And during this time no progress is made. But if the worker is suspended without taking ownership of the lock, then some other worker might be able to acquire synchronously the lock, without incurring the penalty of thread-switching. This worker will be most likely the same worker that currently owns the lock.

The performance comparison that I mentioned previously is with `=> ValueTask.CompletedTask` as workload.

**Comment by stephentoub** [comment]
> After awaiting the semaphore.WaitAsync(), the semaphore has been acquired, but the worker who owns it must resume asynchronously on a different thread

What do you mean?  If WaitAsync synchronously acquires the lock, it'll return a synchronously completed Task, and the await on it won't queue anything.

**Comment by theodorzoulias** [comment]
Yes, the first worker will acquire synchronously the lock with the `WaitAsync`, but after the `Release` this worker won't be able to retake the lock immediately, because now there is a queue of other workers who have requested the lock, and will acquire it in FIFO order. Maintaining the FIFO order is not important in this scenario. It is more important to hand the lock to whoever worker is able to use it right now, without going through a thread-switch process.

**Comment by stephentoub** [comment]
I'm still not understanding the difference what difference you're trying to describe.  Can you just share the code for your experiment?  Then I can just read the code.

**Comment by theodorzoulias** [comment]
Sure. Currently it's a bit messy, so I'll clean it a bit and I'll post the link later.

**Comment by theodorzoulias** [comment]
My initial experiment with the idea of suspending workers, was with a custom locking mechanism `ParallelAsyncLock` that doesn't have thread-affinity, and so it could be used for synchronizing the enumeration of asynchronous sequences as well. The same idea with the `Monitor` as locker is even faster: it's only three times slower than using a simple `lock`. I update [the demo](https://dotnetfiddle.net/9lxoR1), replacing the initial implementation with the faster one, and I got this updated measurement:

-- await ForEachAsync_Suspend(source,
Duration: 1,492 msec
Allocated: 24,200 bytes
LockContentionCount: 64

Comparing the "Suspend" and "Discard" ideas, I am thinking that they are essentially the same idea, with the difference that the former tries to do the work of the garbage collector at managing memory. Which of course the garbage collector can do better, hence discarding workers is faster than suspending them.

**Comment by stephentoub** [comment]
I still don't see the fundamental difference.  With a SemaphoreSlim.WaitAsync, the first acquirer synchronously gets ownership of the lock.  In the meantime, other workers try to acquire the lock and end up adding themselves to a queue.  When the owner of the lock releases it, a single waiter is transferred ownership of the lock and is queued to continue processing as the previous owner proceeds to process the item they retrieved.  That's exactly what happens in your "suspend" implementation as well.

The perf difference I see is just whether using SemaphoreSlim.WaitAsync, which will allocate a new Task on each use, versus using a custom lock that's able to pool IValueTaskSource implementations.  That was never up for debate.

So, what am I missing?

**Comment by theodorzoulias** [comment]
> When the owner of the lock releases it, a single waiter is transferred ownership of the lock and is queued to continue processing as the previous owner proceeds to process the item they retrieved.

That's the bit that slows the processing down: the new owner of the lock is queued. In the "suspend" implementation a suspended worker is not resumed having the ownership of the lock. It is resumed without the ownership. It will try again to take the lock, and may fail again to take it, and as a result will be suspended again without doing any work. Which is desirable in the case that the work is so lightweight, as in our `=> ValueTask.CompletedTask` scenario, that a single worker can do it faster than multiple workers.

> The perf difference I see is just whether using SemaphoreSlim.WaitAsync, which will allocate a new Task on each use, versus using a custom lock that's able to pool IValueTaskSource implementations.

That was also my initial assumption, but my experiments showed no measurable difference between a `SemaphoreSlim` and a `IValueTaskSource`-based custom locker. It seems that the cost of the `Task` allocation is balanced out by the cost of resetting the somewhat heavier `ManualResetValueTaskSourceCore<T>`.

Btw the performance of the `Parallel.ForEachAsync` with asynchronous `source` could be improved by omitting to pass the cancellation token in [the line](https://github.com/dotnet/runtime/blob/65f5ad06c38fbd0a73583a46889419068534446a/src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs#L256) `await state.Lock.WaitAsync`, and instead checking again the `state.Cancellation.IsCancellationRequested` after the `await`. This way the loop with `=> ValueTask.CompletedTask` payload is about 20% faster, and the allocations are significantly reduced.

**Comment by theodorzoulias** [comment]
A small performance improvement. After replacing these lines:

```C#
object locker = new();
//...
if (Monitor.TryEnter(locker))
//...
finally { Monitor.Exit(locker); }
```

...with those:


```C#
int mutex = 0;
//...
if (Interlocked.CompareExchange(ref mutex, 1, 0) == 0)
//...
finally { Volatile.Write(ref mutex, 0); }
```

... the performance of both the "Discard" and "Suspend" ideas was slightly improved. I updated again [the demo](https://dotnetfiddle.net/9lxoR1), and I got these measurements:

```
-- await ForEachAsync_Discard(source,
Duration: 688 msec
Allocated: 307,624,544 bytes
LockContentionCount: 1

-- await ForEachAsync_Suspend(source,
Duration: 1,285 msec
Allocated: 24,168 bytes
LockContentionCount: 130
```

This is interesting because the `Interlocked`-based locking is not thread-affine, and so it could be used for asynchronous `source` sequences as well.

**Comment by stephentoub** [comment]
> In the "suspend" implementation a suspended worker is not resumed having the ownership of the lock

I see. That's a double-edged sword. You force more contention onto the lock in doing so, while also putting more pressure onto the pool unmecessarily. It's tuning to a specific benchmark.

**Comment by stephentoub** [comment]
> the performance of both the "Discard" and "Suspend" ideas was slightly improved

Using a store release fence instead of a full barrier has pros and cons. It is going to be cheaper on x86/64 where there's a strong memory model; your numbers would look different on arm where there's real cost to the fence. It also allows work to move into the lock, such that it could increase contention wherever. It's why Monitor uses a full fence for exit, and why SpinLock uses a full fence by default.

**Comment by stephentoub** [comment]
> Btw the performance of the Parallel.ForEachAsync with asynchronous source could be improved by omitting to pass the cancellation token in [the line](https://github.com/dotnet/runtime/blob/65f5ad06c38fbd0a73583a46889419068534446a/src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs#L256) await state.Lock.WaitAsync, and instead checking again the state.Cancellation.IsCancellationRequested after the await.

Then anyone waiting for the lock will not wake up until its able to be aquired.

**Comment by theodorzoulias** [comment]
> I see. That's a double-edged sword. You force more contention onto the lock in doing so, while also putting more pressure onto the pool unnecessarily. It's tuning to a specific benchmark.

The goal is to fulfil the promise of asynchronous programming, which is the efficient use of threads, even in the rare scenario of a `source` sequence with blocking behavior, without hurting too much the performance in the equally rare scenario of using the  `Parallel.ForEachAsync` with synchronous lightweight payload. Switching to a benchmark with a real-life payload, like downloading a web page from a remote web server, is likely to eliminate all contention for accessing the enumerator, since the enumeration work will be totally dwarfed by the download work.

> Using a store release fence instead of a full barrier has pros and cons. It is going to be cheaper on x86/64 where there's a strong memory model; your numbers would look different on arm where there's real cost to the fence. It also allows subsequent work to move into the lock, such that it could increase contention elsewhere. It's why Monitor uses a full fence for exit, and why SpinLock uses a full fence by default.

Interesting. Lock-free multithreading is not my expertise, so doing mistakes is expected. I switched from `Volatile.Write` to `Interlocked.Exchange` in [the demo](https://dotnetfiddle.net/9lxoR1), and the performance got worse by a tiny bit, but it's still better than using the `Monitor`. Which was kind of surprising to be honest.

> Then anyone waiting for the lock will not wake up until its able to be acquired. Is your thinking it's not necessary because it'll only be blocked by other operations in the same iteration and thus will be unblocked anyway when they're canceled and release the lock?

Yes, exactly!
